### DISCUSSION TITLE
Compute, Communication, and Overlap Time

### DISCUSSION THREAD

**Discussion by aruhela** [horovod/horovod#2792] [Q&A]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

Is it possible to calculate programmatically the time spent in Compute, Communication, and their overlap from the horovod timeline? How accurate is the time compared to other profilers like nvprof and mpiP?

Has anyone created a script to determine these timings?

Thanks,
Amit

**Comment by tgaddair** [comment]
Hey @aruhela, you may want to check out #2723, which adds some profiling features that may provide what you're looking for.

@romerojosh do you have any additional tips for profiling Horovod overlap of compute / communication?

**Reply by romerojosh** [reply]
As alluded to by @tgaddair, I use the nsight-systems profiler from NVIDIA to get a visual idea of how well my compute and communication are overlapping. Horovod operations are launched into their own stream so it is usually fairly straightforward to find the stream with NCCL and Horovod operations in it, and see when/how those kernels are overlapping with compute kernels launched by the framework in the one or more other streams. #2723 which is now merged into master should provide some nice NVTX annotations into the profile to get a better idea of what Horovod is doing.

As far as programmatically determining these times, I do not have any general advice or a script to do this. I actually very rarely try to compute these specific timings since I think a more general measurement of the performance tends to be more useful in these situations. What I mean by this is I will do something like the following:
1. Measure performance (iteration times, img/sec, etc) of a run using single GPU.
2. Measure performance of a run using multiple GPUs _without_ Horovod enabled to gauge any limitations caused by limitations of shared on-node resources (e.g. I/O might be able to keep up with a single GPU, but might slow down when multiple GPUs on the same node are sharing resources)
3. Measure performance of a run using multiple GPUs _with_ Horovod enabled.

Comparing the performance you see in (3) relative to (1) or (2) will give you an idea of how much the communication in Horovod costs and using those metrics, you can adjust Horovod parameters to see if you can reduce it.

**Reply by ihchoi12** [reply]
@romerojosh Sorry for asking a silly question. How can I measure the performance with multiple GPUS without Horovod enabled? I thought the Horovod is enabling the communication across multiple GPUs. Can we achieve the communication without horovod?

**Comment by aruhela** [comment]
Thanks Josh and Travis,

Many a time, we need to plot the trend or compare the communication time which requires the numerical values of communication times for the entire run. Using different communication algorithms/MPI/NCCL, we want to demonstrate improvements in communication time for multi-node runs.   I believe this should be doable by scanning the horovod timelines in a bash or python script .
